Added extended rules to be able to plugin BUILDER_PLUGINS configurations

nrgaway opened the pull request:
- If a BUILDER_PLUGINS is defined, it will also load its COMPONENT to prevent
  the need to define both
- BUILDER_PLUGINS_DISTS are also parsed and loaded
- If a builder.conf exists for the PLUGIN, it is imported
- Added formatting and color to `build-info` target
    - Only displayed when VERBOSE > 0 to allow better overview of what
      components, flavors, etc are loaded.
    - The info will show anything that was modified from the original values
      within configuration files in red (automatic aliases, labels, etc)
    - Any items removed by configuration will show as white

nrgaway commented:
How this works...

define BUILDER_PLUGINS += app-salt-config

extended_rules.conf will auto-load the COMPONENT `app-salt-config` only if it is defined in `BUILDER_PLUGINS` or `BUILDER_PLUGINS_DIST`.  Then if the `PLUGIN` has a `builder.conf` file in its base directory, it will be included.

I use the `app-salt-config` `builder.conf` to add the `salt` COMPONENT add the `salt` flavor to either manually selected DIST_VM (by adding +salt), or by pattern base on `TEMPLATE_FLAVOR_SALT` var (where that can contain fc, fc20, whonix or some pattern to match which will be applied`.  The COMPONENTS get unloaded if no flavor is selected and the `BUILDER_PLUGIN` also gets unloaded.

I created the color coded makefile details to visually be able to see whats happening with all the vars (based on `VERBOSE` levels) and formatted it somewhat to make it pretty :)

See https://github.com/nrgaway/qubes-app-salt-config/blob/develop-wip/builder.conf for example usage

marmarek commented:
Generally I'm ok with this change, but I'm not convinced that
*example-configs*/extended-rules.conf is the best place for such things.
This directory should be for... example configs, not a builder
implementation. qubes-os*.conf and templates.conf are some examples what
should be there.

In the current layout, with changes from this pull request accepted,
extended-rules.conf should be:
* moved to builder.conf in builder plugins - distro-specific parts
  (especially whonix-specific rules -> builder-debian)
* merged into main Makefile (distro-independed parts)

This will make the main Makefile somehow long, so I'm thinking of
splitting it into some separate files, but this is irrelevant here.

Some unsorted comments about extended-rules.conf:
* what is `template-modules` for? How is it different from "qubes-vm"
  (besides running some whonix-specific targets first)?
* plugins shouldn't be automatically enabled based on DISTS_VM - this
  thing should be explicitly enabled in the config file. The same applies to
  other components (like whonix components) - what code is downloaded
  and included in the build process should be a conscious user decision;
  builder-debian plugin can provide convenient setting
  (`WHONIX_COMPONENTS`), but the only place where main `COMPONENTS`
  setting is modified should main builder.conf. `setup` can provide an
  easy way to generate such config file, also `example-configs` is a
  good place for such examples.


-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

nrgaway commented:
Well, good news then... I just finished re-factoring `Whonix` into its very own package which enabled me to rip out Whonix specific code out of `extended-rules.conf` and `builder-debian` using the `BUILDER_PLUGIN` API in this commit.  

I also moved the template code that was in `templates_debian` into the new `whonix` package.  Its named `qubes-template-whonix`, but I can rename it to have the word `flavor` or maybe you have a better idea.

Currently though, you may need to re-run `get-sources` 2 or 3 times to get all the source files.  Maybe I should look at `COMPONENTS` before get-sources is run and compare it afterwards to see if any addition `COMPONENTS` were loaded and just grab them in the same run.

`template-modules` allow to only download and build the template-modules and run execute a few additional targets such as umount (since that was always giving me issues when build fail), patch can be moved. build-info gives a nice overview of what is being loaded in regards to the overall configuration options, which helps me debug and confirm the correct variables are being set.  I show configurations that were modified, either added or removed, as well (alias will change name of dist, removal of unused template labels so they are not looped though, removal of BUILDER_PLUGIN (I remove salt if not being used to prevent building).

I will rip the code out that enables any BUILDER_PLUGINS and place at top of main files as suggested, and give an option in setup to enable them as well.  Currently in the Whonix package I did enable the `builder-debian` plugin if it was not enabled, but I could replace that with a build error informing the user it also needs to be enabled.

On the note of BUILDER_PLUGINS, do I need to enable one to build the ISO?  I have not been able to build ISO for about a week, always using your most current repo (Trying to test the salt module).  I don't have the error handy since I am running whonix test builds, but it pretty much fails straight away stating something like rpm cant be found (installer package error).

marmarek commented:
On Sat, Mar 28, 2015 at 04:11:27PM -0700, nrgaway wrote:
> Currently though, you may need to re-run `get-sources` 2 or 3 times to get all the source files.  Maybe I should look at `COMPONENTS` before get-sources is run and compare it afterwards to see if any addition `COMPONENTS` were loaded and just grab them in the same run.

Yes, this can be somehow annoying. Currently code to build the template
in DispVM does the same:
        make get-sources COMPONENTS='builder $(BUILDER_PLUGINS)';
        make get-sources;

> I will rip the code out that enables any BUILDER_PLUGINS and place at top of main files as suggested, and give an option in setup to enable them as well.  Currently in the Whonix package I did enable the `builder-debian` plugin if it was not enabled, but I could replace that with a build error informing the user it also needs to be enabled.

I think the build error would be better.

> On the note of BUILDER_PLUGINS, do I need to enable one to build the ISO?  I have not been able to build ISO for about a week, always using your most current repo (Trying to test the salt module).  I don't have the error handy since I am running whonix test builds, but it pretty much fails straight away stating something like rpm cant be found (installer package error).

You need at least builder-fedora enabled. I've update
example-configs/qubes-os-master.conf to reflect latest changes.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

marmarek commented:
Are the changes you've mentioned available anywhere?

nrgaway commented:
On 7 April 2015 at 17:47, Marek Marczykowski-Górecki <
notifications@github.com> wrote:

> Are the changes you've mentioned available anywhere?
>

Yes, I will push them in a few hours; just completing final testing

—
> Reply to this email directly or view it on GitHub
> <https://github.com/marmarek/qubes-builder/pull/6#issuecomment-90741947>.
>

nrgaway commented:
Here are the changes.  The commit log messages are fairly descriptive, but a very quick overview...

- Removed extended-rules.conf and moved relevant rules/targets to Makefile or to specific builder configuration files.
- Added the `LOADING_PLUGINS` var to Makefile.generic as discussed above
- Converted setup to use python as its way more flexible; created new format for setup configuration file (`.setup.data`)
- Setup now only uses one configuration file instead of the 3 or 4 as requested
- Setup's template configuration file now has all BUILDER_PLUGINS disabled by default; need to enabled them via setup of manually within builder.conf
- All of Whonix configurations and templates are now within their own package (qubes-template-whonix) and implemented as a BUILDER_PLUGIN (https://github.com/nrgaway/qubes-template-whonix)
